telegrambots-abilities/src/main/java/org/telegram/abilitybots/api/objects/Ability.java

```diff
@@ -63,7 +67,15 @@ private Ability(String name, String info, Locality locality, Privacy privacy, in
         "Use the number 0 if the method ignores the arguments OR uses as many as appended");
     this.argNum = argNum;
 
-    this.action = checkNotNull(action, "Method action can't be empty. Please assign a function by using .action() method");
+    this.webhookAction = webhookAction;
+    boolean isWebhook = webhookAction != null;
+
+    if (!isWebhook) {
+      this.action = checkNotNull(action, "Method action can't be empty. Please assign a function by using .action() method");
+    } else {
+      this.action = null;
+    }
```

Comment:
You're enforcing a webhook action on all declared abilities. I haven't looked at your compilation/testing results, but I guess this is why your Travis build failed.

You should rethink how you're going to make the framework accept an action that returns a method. Perhaps this warrants a WebhookAbility that changes the action in a function.

Reply:
That also has other implications since Ability is final. Re-architecting that needs to be handled with care.

Reply:
"You're enforcing a webhook action on all declared abilities."
I don't get what you mean by this.

Reply:
Sorry, I misread. No, we're not enforcing here the webhook action, but I think this may allow users of the long-polling variant to assign a webhook action without assigning an action and have that pass the  pre-conditions. Another caveat is that developers would be able to declare an ability with both actions, yet only one will be acted on. A third caveat would be that the webhook action is not properly checked for null. Unlike the normal action, this wouldn't throw an exception if the webhook action is null. That's because you're using that exact logic to determine whether the Ability is webhook or not. That's bad design. I think we can do better.

What I'm saying here is, relying on whether the user has supplied a webhook action to determine whether the Ability itself targets a Webhook bot seems like a very loose relationship to me. It would be a stronger implication to declare a Webhook Ability (or even push this logic up to the builder!).

I'm here to help, so if anything is unclear, shoot me a DM.

Reply:
So, we can do something like:
1 - BaseAbility
2 - Ability
3 - WebhookAbility
4 - AbilityBuilder <- abstract and package-private

Then the Ability variants will have their own builders, extending the AbilityBuilder and performing the separate action setting.

Reply:
I wanted to go with that suggestion from the beginning but I didn't want to change that much code, fearing it might break so much stuff along the way.
All issues mentioned above will be fixed automatically if we go with that approach. 

Other than that, I really like the idea and will implement it in a separate branch.
And I will be happy to hear your comments/suggestions on it <3.

Reply:
Any updates on this PR? A new version is coming soon and if you got something down we could get it merged. Otherwise, it'll have to wait for the next cycle.

Reply:
Sorry, I've been too busy.
we'd have to wait for the next cycle.